diff -uNr gst-plugins-good-1.22.9.a/sys/v4l2/gstv4l2videoenc.c gst-plugins-good-1.22.9.b/sys/v4l2/gstv4l2videoenc.c
--- gst-plugins-good-1.22.9.a/sys/v4l2/gstv4l2videoenc.c	2024-09-27 16:38:10.409814540 +0800
+++ gst-plugins-good-1.22.9.b/sys/v4l2/gstv4l2videoenc.c	2024-10-10 16:35:41.449283882 +0800
@@ -553,13 +553,6 @@
   GST_DEBUG_OBJECT (self, "Negotiating %s profile and level.",
       klass->codec_name);
 
-  if(klass->set_option){
-    if (!klass->set_option (self)) {
-      GST_ERROR_OBJECT (self, "subclass failed to set option");
-      return FALSE;
-    }
-  }	  
-
   /* Only renegotiate on upstream changes */
   if (self->input_state)
     return TRUE;
@@ -567,6 +560,13 @@
   if (!codec)
     goto done;
 
+  if(klass->set_option){
+    if (!klass->set_option (self)) {
+      GST_ERROR_OBJECT (self, "subclass failed to set option");
+      return FALSE;
+    }
+  }
+
   allowed_caps = gst_pad_get_allowed_caps (GST_VIDEO_ENCODER_SRC_PAD (encoder));
 
   if (allowed_caps) {
@@ -689,6 +689,18 @@
     if (cpool)
       gst_object_unref (cpool);
   }
+  /* Wait stream on if get last buffer */
+  if(ret == GST_V4L2_FLOW_LAST_BUFFER || ret == GST_FLOW_EOS){
+    if(self->wait_streamon_times){
+      g_usleep(10*1000);
+      self->wait_streamon_times--;
+      ret = GST_FLOW_OK;
+    }
+  }
+  else{
+    self->wait_streamon_times = 15;
+  }
+
   if (ret != GST_FLOW_OK)
     goto beach;
 
@@ -902,6 +914,11 @@
 
     {
       GstBufferPool *opool = gst_v4l2_object_get_buffer_pool (self->v4l2output);
+
+      if(!gst_buffer_pool_is_active(opool))
+        if(task_state == GST_TASK_STARTED)
+          gst_buffer_pool_set_active(opool, TRUE);
+
       ret = gst_v4l2_buffer_pool_process (GST_V4L2_BUFFER_POOL (opool),
           &frame->input_buffer, &frame->system_frame_number);
       if (opool)
@@ -1167,6 +1184,7 @@
       gst_v4l2_get_input, gst_v4l2_set_input, NULL);
 
   self->frame_num = -1;
+  self->wait_streamon_times = 15;
 }
 
 static void
diff -uNr gst-plugins-good-1.22.9.a/sys/v4l2/gstv4l2videoenc.h gst-plugins-good-1.22.9.b/sys/v4l2/gstv4l2videoenc.h
--- gst-plugins-good-1.22.9.a/sys/v4l2/gstv4l2videoenc.h	2024-09-27 16:38:10.409814540 +0800
+++ gst-plugins-good-1.22.9.b/sys/v4l2/gstv4l2videoenc.h	2024-09-27 16:40:47.605892009 +0800
@@ -68,7 +68,7 @@
 
   /**/
   gint64 frame_num;
-
+  gint64 wait_streamon_times;
 };
 
 struct _GstV4l2VideoEncClass
