diff -uNr gst-plugins-bad-1.22.9/gst/videoparsers/gsth264parse.c gst-plugins-bad-1.22.9.m/gst/videoparsers/gsth264parse.c
--- gst-plugins-bad-1.22.9/gst/videoparsers/gsth264parse.c	2024-01-25 02:21:13.000000000 +0800
+++ gst-plugins-bad-1.22.9.m/gst/videoparsers/gsth264parse.c	2024-07-02 18:09:06.135579838 +0800
@@ -197,8 +197,8 @@
 gst_h264_parse_init (GstH264Parse * h264parse)
 {
   h264parse->frame_out = gst_adapter_new ();
-  gst_base_parse_set_pts_interpolation (GST_BASE_PARSE (h264parse), FALSE);
-  gst_base_parse_set_infer_ts (GST_BASE_PARSE (h264parse), FALSE);
+  gst_base_parse_set_pts_interpolation (GST_BASE_PARSE (h264parse), TRUE);
+  gst_base_parse_set_infer_ts (GST_BASE_PARSE (h264parse), TRUE);
   GST_PAD_SET_ACCEPT_INTERSECT (GST_BASE_PARSE_SINK_PAD (h264parse));
   GST_PAD_SET_ACCEPT_TEMPLATE (GST_BASE_PARSE_SINK_PAD (h264parse));
 
